patches/0003-Add-crypto-backend-foundation.patch

```diff
@@ -83,6 +85,73 @@ index 097c37e343fdb8..1cf43edba40359 100644
  
  // Enabled reports whether BoringCrypto handles supported crypto operations.
  func Enabled() bool {
+diff --git a/src/crypto/ecdh/ecdh.go b/src/crypto/ecdh/ecdh.go
+index e5270d840b9abf..29d8665b809f2a 100644
+--- a/src/crypto/ecdh/ecdh.go
++++ b/src/crypto/ecdh/ecdh.go
+@@ -8,7 +8,7 @@ package ecdh
+ 
+ import (
+ 	"crypto"
+-	"crypto/internal/boring"
++	boring "crypto/internal/backend"
+ 	"crypto/subtle"
+ 	"io"
+ 	"sync"
+diff --git a/src/crypto/ecdh/nist.go b/src/crypto/ecdh/nist.go
+index 01354fa2cf0af2..7fa4c338a168ee 100644
+--- a/src/crypto/ecdh/nist.go
++++ b/src/crypto/ecdh/nist.go
+@@ -5,11 +5,12 @@
+ package ecdh
+ 
+ import (
+-	"crypto/internal/boring"
++	boring "crypto/internal/backend"
+ 	"crypto/internal/nistec"
+ 	"crypto/internal/randutil"
+ 	"encoding/binary"
+ 	"errors"
++	"internal/goexperiment"
+ 	"io"
+ 	"math/bits"
+ )
+@@ -36,7 +37,7 @@ func (c *nistCurve[Point]) String() string {
+ var errInvalidPrivateKey = errors.New("crypto/ecdh: invalid private key")
+ 
+ func (c *nistCurve[Point]) GenerateKey(rand io.Reader) (*PrivateKey, error) {
+-	if boring.Enabled && rand == boring.RandReader {
++	if boring.Enabled && rand == boring.RandReader && !goexperiment.OpenSSLCrypto {
+ 		key, bytes, err := boring.GenerateKeyECDH(c.name)
+ 		if err != nil {
+ 			return nil, err
+@@ -79,7 +80,7 @@ func (c *nistCurve[Point]) NewPrivateKey(key []byte) (*PrivateKey, error) {
+ 	if isZero(key) || !isLess(key, c.scalarOrder) {
+ 		return nil, errInvalidPrivateKey
+ 	}
+-	if boring.Enabled {
++	if boring.Enabled && !goexperiment.OpenSSLCrypto {
+ 		bk, err := boring.NewPrivateKeyECDH(c.name, key)
+ 		if err != nil {
+ 			return nil, err
+@@ -173,7 +174,7 @@ func (c *nistCurve[Point]) NewPublicKey(key []byte) (*PublicKey, error) {
+ 		curve:     c,
+ 		publicKey: append([]byte{}, key...),
+ 	}
+-	if boring.Enabled {
++	if boring.Enabled && !goexperiment.OpenSSLCrypto {
+ 		bk, err := boring.NewPublicKeyECDH(c.name, k.publicKey)
+ 		if err != nil {
+ 			return nil, err
+@@ -196,7 +197,7 @@ func (c *nistCurve[Point]) ecdh(local *PrivateKey, remote *PublicKey) ([]byte, e
+ 	// only be the result of a scalar multiplication if one of the inputs is the
+ 	// zero scalar or the point at infinity.
+ 
+-	if boring.Enabled {
++	if boring.Enabled && !goexperiment.OpenSSLCrypto {
```

Comment:
Shouldn't this `goexperiment.OpenSSLCrypto` usage be in the OpenSSL patch rather than the foundation patch?

Reply:
Yep, good catch.

Reply:
Fixed!